akka-stream/src/main/scala/akka/stream/impl/SynchronousFilePublisher.scala

```diff
@@ -0,0 +1,117 @@
+/**
+ * Copyright (C) 2015 Typesafe Inc. <http://www.typesafe.com>
+ */
+package akka.stream.impl
+
+import java.io.{ File, RandomAccessFile }
+import java.nio.channels.FileChannel
+
+import akka.actor.{ DeadLetterSuppression, ActorLogging, Props }
+import akka.io.DirectByteBufferPool
+import akka.stream.actor.ActorPublisherMessage
+import akka.util.ByteString
+
+/** INTERNAL API */
+private[akka] object SynchronousFilePublisher {
+  def props(f: File, chunkSize: Int, initialBuffer: Int, maxBuffer: Int) = {
+    require(chunkSize > 0, s"chunkSize must be > 0 (was $chunkSize)")
+    require(initialBuffer > 0, s"initialBuffer must be > 0 (was $initialBuffer)")
+    require(maxBuffer >= initialBuffer, s"maxBuffer must be >= initialBuffer (was $maxBuffer)")
+
+    Props(classOf[SynchronousFilePublisher], f, chunkSize, initialBuffer, maxBuffer)
+  }
+
+  private final case object Continue extends DeadLetterSuppression
+}
+
+/** INTERNAL API */
+private[akka] class SynchronousFilePublisher(f: File, chunkSize: Int, initialBuffer: Int, maxBuffer: Int) extends akka.stream.actor.ActorPublisher[ByteString]
+  with ActorLogging {
+
+  import SynchronousFilePublisher._
+
+  // TODO also have a Promise[Long] here? for bytes-read?
+
+  val buffs = new DirectByteBufferPool(chunkSize, maxBuffer)
+  var eofReachedAtOffset = Long.MinValue
+
+  var readBytesTotal = 0L
+  var availableChunks: Vector[ByteString] = Vector.empty
+
+  private val fs = new RandomAccessFile(f, "r") // best way to express this in JDK6, OpenOption are available since JDK7
+  private var chan: FileChannel = _
+
+  override def preStart() = {
+    try {
+      chan = fs.getChannel
+    } catch {
+      case ex: Exception ⇒
+        log.error(ex, "Unable to open channel for {}", f.getAbsolutePath)
+        context.stop(self)
```

Comment:
I think this is too early. Should emit error downstreams when `Request` is received?